src/storage/src/upsert.rs

```diff
@@ -349,6 +351,64 @@ where
     }
 }
 
+/// Renders an operator that discards updates that are known to not affect the outcome of upsert in
+/// a streaming fashion. For each distinct (key, time) in the input it emits the value with the
+/// highest from_time. Its purpose is to thin out data as much as possible before exchanging them
+/// across workers.
+fn upsert_thinning<G, K, V, FromTime>(
+    input: &Collection<G, (K, V, FromTime), Diff>,
+) -> Collection<G, (K, V, FromTime), Diff>
+where
+    G: Scope,
+    G::Timestamp: TotalOrder,
+    K: timely::Data + Eq + Ord,
+    V: timely::Data,
+    FromTime: Timestamp,
+{
+    input
+        .inner
+        .unary(Pipeline, "UpsertThinning", |_, _| {
+            // A capability suitable to emit all updates in `updates`, if any.
+            let mut capability: Option<InputCapability<G::Timestamp>> = None;
+            // A batch of received updates
+            let mut updates = Vec::new();
+            let mut tmp = Vec::new();
+            move |input, output| {
+                while let Some((cap, data)) = input.next() {
+                    data.swap(&mut tmp);
+                    updates.append(&mut tmp);
+                    match capability.as_mut() {
+                        Some(capability) => {
+                            if cap.time() <= capability.time() {
+                                *capability = cap;
+                            }
+                        }
+                        None => capability = Some(cap),
+                    }
+                }
+                if let Some(capability) = capability.take() {
+                    // Sort by (key, time, Reverse(from_time)) so that dedupping by (key, time) gives
+                    // the latest change for that key.
+                    updates.sort_unstable_by(|a, b| {
+                        let ((key1, _, from_time1), time1, _) = a;
+                        let ((key2, _, from_time2), time2, _) = b;
+                        Ord::cmp(
+                            &(key1, time1, Reverse(from_time1)),
+                            &(key2, time2, Reverse(from_time2)),
+                        )
+                    });
+                    let mut session = output.session(&capability);
+                    session.give_iterator(updates.drain(..).dedup_by(|a, b| {
+                        let ((key1, _, _), time1, _) = a;
+                        let ((key2, _, _), time2, _) = b;
```

Comment:
do we not care about the `diff` here? Or is there some guarantee that this collection always contains rows with diffs of `+1`?

Reply:
The latter, upsert only works with positive diffs. We assert this fact in `stage_input`

Reply:
I will assert it here too since this function might drop a record that had a negative diff